### PR TITLE
Add resampler

### PR DESCRIPTION
**Pull request by phillip-wenig-frequenz**
Adds a resampler that resamples a single stream of data based on a resampling function in a given interval.

### CODE REVIEW

File: Cargo.toml
Comment (llucax) [inline Cargo.toml]: Following what [we talked about](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/pull/1#discussion_r1787637137) in the [component-graph-rs](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs) lib, I suggest removing the `-rs` suffix from the package name. I'm also fine to keep it if it is common in rust, but I would pick one scheme and use the same in all projects.

File: Cargo.toml
Comment (llucax) [inline Cargo.toml]: Same here, but for the module name I would definitely avoid the `_rs` suffix.

File: README.md
Comment (llucax) [inline README.md]: Maybe add the expected result as a comment or something, in a glance it is not clear to me if you'll get one sample for the first second or what exactly, specially compared with the batch resampling it is not clear to me what's the difference.

File: README.md
Comment (llucax) [inline README.md]: Now I'm interested in knowing how many orders of magnitude faster this is compared to the python implementation :laughing:

File: README.md
Comment (llucax) [inline README.md]: Also out of curiosity it might be nice to include one output of this benchmark, specifying the hardware it was ran on, to have a feeling again at least about the order of magnitude this works on.

File: README.md
Comment (llucax) [inline README.md]: In the SDK we can align to any timestamp, defaulting to epoch. I guess the same will be done here, but mentioning just in case.

File: README.md
Comment (llucax) [inline README.md]: Not sure what these 2 mean, but in any case I would add these as GitHub issues rather than keeping a TODO in the README.

**Comment by phillip-wenig-frequenz** [comment]
@cwasicki @shsms new version pushed:

# mode 1

- has buffer that is filled by a push method
- resample trigger performs resample function on the buffer and empties it
- interval will be handled by the outside logic

# mode 2

- interval size is defined during construction
- does not use buffer, but expects a vector of input streams
- interval is aligned to epoch and input data is resampled, considering gaps, Nones, and resample functions
- no state update during resampling

**Review by llucax — COMMENTED** [review]
I just glanced over the code as rust is not my strong point,, so mostly comments about structure and docs.

File: README.md
Comment (phillip-wenig-frequenz) [reply README.md]: oh yes, I'll remove it

File: README.md
Comment (phillip-wenig-frequenz) [reply README.md]: Is this required to align to any timestamp?

File: README.md
Comment (cwasicki) [reply README.md]: At least for 10k-1M samples pandas is somewhat comparable in my poor-man benchmarks. Slower for smaller sizes (interestingly 1k is consistently slower than 10k) and also not linear beyond that.
```
1000 rows: 1.024 ms (1024 ns per row)
10000 rows: 0.986 ms (99 ns per row)
100000 rows: 3.912 ms (39 ns per row)
1000000 rows: 52.436 ms (52 ns per row)
10000000 rows: 1056.000 ms (106 ns per row)
100000000 rows: 16904.371 ms (169 ns per row)
```

Determined with:
```python
for e in range(3, 9):    
    size = 10**e    
    df = pd.DataFrame({    
        "timestamp": pd.date_range(start=pd.Timestamp.now(), periods=size, freq='s'),    
        "value_0": np.random.rand(size),    
        "value_1": np.random.rand(size)    
    })    
    df.set_index('timestamp', inplace=True)    
    start_time = time.time()    
    resampled_df = df.resample('5s').asfreq()    
    end_time = time.time()    
    duration = end_time - start_time    
    print(f"{size} rows: {1e3 * duration:.3f} ms ({1e9 * duration / size:.0f} ns per row)")  
```

File: README.md
Comment (cwasicki) [reply README.md]: Do we have a use-case for aligning to any timestamp?

File: README.md
Comment (phillip-wenig-frequenz) [reply README.md]: I am looking into speeding it up a bit. Is multithreading an option in the environment it will run in? @llucax @cwasicki

Comment (cwasicki) [reply README.md]: At least in my opinion we don't need to worry about this now. Just wanted to argue a bit against the bad reputation pandas has :) In fact the comparison is not completely fair (e.g. due to data structure) and may be easier to optimize than in our case.

Comment (phillip-wenig-frequenz) [reply README.md]: Alright, made it 2x faster 👯

Comment (phillip-wenig-frequenz) [reply README.md]: ...without multithreading

File: README.md
Comment (llucax) [reply README.md]: Not sure, I don't remember now, there was a lot of discussion about different use cases. But once you do the math to align to the epoch using any other arbitrary point in time is for free, so not sure why not supporting it?

File: README.md
Comment (llucax) [reply README.md]: I think as long as threads are internal to the library it should be fine. Not sure if there are any guidelines in place for rust in particular @frequenz-floss/api-dispatch-team

File: README.md
Comment (llucax) [inline README.md]: :100:

**Comment by shsms** [comment]
I think having to track `stream_id` might be harder for users to deal with than having one `Resampler` instance per stream.

I think we should also support only one stream per instance, because then we can dynamically start and stop resamplers for specific metrics as necessary, without having to know which metrics to process at startup.

This also makes it easier for outer layers to decide if they want to run different resamplers in different threads, etc.

I was thinking the driving for the historical data can also come from the outside.  After this is merged, we can design a historical data driver that fills the buffers and calls resample at every window.  Then this class can become even simpler - just deal with the buffer for the window.

Btw, the `Resampler` also needs support holding onto data for longer, like described here: https://github.com/frequenz-floss/frequenz-sdk-python/blob/443e83e599be508d832beec13884d3e5f18e6c35/src/frequenz/sdk/timeseries/_resampling.py#L140-L160

So we only clear data from the buffer that's older the `max_data_age * resampling_interval`.  This has to be the case for both live and historical data.

But maybe we can get this code to a state where it doesn't have any special code for live or historical data.

EDIT: Also I guess for this you'd need to take the timestamp of the next message from the outside, along with the trigger to resample.

File: src/resampler.rs
Comment (shsms) [inline src/resampler.rs]: I'd think `Box<dyn FnMut...>` is the most flexible.  It would mean we can't `Clone` anymore, but do we really need clone?

`ResamplingFunction` will be a member of the `Resampler`, and there will always be a mutable reference to call `Box::as_mut`, before you apply the function.

Comment (shsms) [reply src/resampler.rs]: `FnMut` is more flexible because you can also include immutable functions. But with `Fn`, you can only specify callables that don't have a mutable state. I think mutable states would be helpful for custom resampler functions.

File: src/resampler.rs
Comment (shsms) [inline src/resampler.rs]: You were going to make this accept a generic iterator just like `Vec::extend` does, right?

File: src/resampler.rs
Comment (shsms) [inline src/resampler.rs]: We already have a lot of code that does it like this, so just for consistency:
```suggestion
// License: MIT
// Copyright © 2024 Frequenz Energy-as-a-Service GmbH
```

File: src/lib.rs
Comment (shsms) [inline src/lib.rs]:
```suggestion
// License: MIT
// Copyright © 2024 Frequenz Energy-as-a-Service GmbH
```

File: src/resampler.rs
Comment (tiyash-basu-frequenz) [inline src/resampler.rs]: I think this can be auto derived using `#[derive(Default)]` by using the `#[default]` attribute right above the variant `Average`.

File: src/resampler.rs
Comment (tiyash-basu-frequenz) [inline src/resampler.rs]: Would be nice to document what happens if it is `None`.

File: src/resampler.rs
Comment (shsms) [inline src/resampler.rs]: I just noticed this when reading the docs: I think we should make `max_age` and `start_ts` required arguments rather than having functions that can update them.

This is a low-level library, so it would be nicer to make the interface more explicit and not have default values. This would also ensure that people don't forget to call the set methods or call them after the resampler already has an internal state.

**Review by tiyash-basu-frequenz — COMMENTED** [review]
Looks pretty neat! I have a few nits which can be ignored.

File: src/resampler.rs
Comment (shsms) [inline src/resampler.rs]: This should be `buffer` to be idiomatic: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

File: src/lib.rs
Comment (shsms) [inline src/lib.rs]: I think because constructors don't have standardized names, we should specify the function's name.  Plus you could also link to specific functions under the Usage section like this:

```suggestion
An instance of the [`Resampler`] can be created with the [`new`][Resampler::new]
method.

Raw data can be added to the resampler either through the
[`push`][Resampler::push] or [`extend`][Resampler::extend] methods, and the the
[`resample`][Resampler::resample] method resamples the data that was added to
the buffer.
```

**Review by shsms — COMMENTED** [review]
Some more, but just incremental things.  As @tiyash-basu-frequenz says, this looks very nice already.

File: src/resampler.rs
Comment (phillip-wenig-frequenz) [reply src/resampler.rs]: Do we need the epoch alignment, if `start` is set by the outside?

Comment (shsms) [reply src/resampler.rs]: Yes, I think so.  Epoch alignment should always be there.   The documentation could clarify that this needs to be `Some` only for historical data.  If this is not specified and the historical data starts with a gap, the resampler can't detect it.

Comment (shsms) [reply src/resampler.rs]: okay, you made it not `Option`al.  That makes sense as well, of course.

File: src/lib.rs
Comment (shsms) [inline src/lib.rs]:
```suggestion
```

File: src/resampler.rs
Comment (shsms) [inline src/resampler.rs]:
```suggestion
    /// Resample the data in the buffer that is not older than max_age. Number of
    /// intervals.
    max_age: i32,
```

File: src/resampler.rs
Comment (shsms) [inline src/resampler.rs]:
```suggestion
    /// The start time of the resampling.
    start: DateTime<Utc>,
```

File: src/resampler.rs
Comment (phillip-wenig-frequenz) [reply src/resampler.rs]: yes, as you said, it's a low level lib and we can define the start from the outside in any case.

File: README.md
Comment (shsms) [inline README.md]: Does this make sense @llucax?  We want `max_age >= 1` right?

Comment (llucax) [reply README.md]: To match the SDK meaning, yes, we just multiply this value by the interval to get the real maximum age the data can have, so using 0 means no data is used at all when resampling.

I also find `max_age` too ambiguous, I would use `max_age_in_intervals` to sort of match what we have in the SDK, but maybe rust style is to use more concise variable names. At the company-level the idea was to try to make the code as obvious as possible so I guess we need to balance between common language style and *company policy*.

BTW, reading the [SDK docs about this variable](https://frequenz-floss.github.io/frequenz-sdk-python/latest/reference/frequenz/sdk/actor/#frequenz.sdk.actor.ResamplerConfig.max_data_age_in_periods) it also reminded me that the resampler needs to be able to deal with down sampling but up sampling too, and in the later case, it doesn't make sense to use the resampling interval for this, but the input sampling interval (if what we want to achieve is to be able to produce data even if samples don't came for a couple of consecutive intervals).

I strongly recommend going through the resampling  code in the SDK to see all the possible ways the resampler could get tricky and bizarre :P

Comment (phillip-wenig-frequenz) [reply README.md]: Yes, we have the functionality here that the interval that is multiplied with `max_age` adapts to the input interval, if the input interval is larger than the set `interval`.

```rust
self.input_interval = Some((s.timestamp() - input_start).max(self.interval));
```

Is that already what you have in mind, @llucax ?

There is also a test for this behavior: [a0262fea108af44a98cc49e5eb72641963dd816513f2d0a22eb738fcfed55ebeR201](https://github.com/frequenz-floss/frequenz-resampling-rs/pull/1/files#diff-a0262fea108af44a98cc49e5eb72641963dd816513f2d0a22eb738fcfed55ebeR201)

File: src/tests.rs
Comment (cwasicki) [inline src/tests.rs]: Check: Is 0 different from 1? If so, would 3 mean using the data from 4 intervals?

Comment (phillip-wenig-frequenz) [reply src/tests.rs]: Yes, that is correct! We resample, then we empty the buffer until `current - interval * max_age`. If `max_age` is `0`, no old values remain and in the next interval we only have values from this interval. If `max_age` is `1`, we have the last interval and the new one. Here are tests that show the behavior: https://github.com/frequenz-floss/frequenz-resampling-rs/pull/1/files#diff-a0262fea108af44a98cc49e5eb72641963dd816513f2d0a22eb738fcfed55ebeR138

**Comment by phillip-wenig-frequenz** [comment]
I have found a small bug regarding `max_age` when we re-call the `resample` function. Is fixed now :)

File: src/tests.rs
Comment (cwasicki) [reply src/tests.rs]: To me this feels not intuitive, because a `max_age` of 1 takes data from 2 sample intervals. It could change its behavior to shift by 1, or the name could change (e.g. `reuse_previous_intervals`, but probably there is a better name).

File: src/lib.rs
Comment (cwasicki) [inline src/lib.rs]: The max age argument is required right?

Comment (phillip-wenig-frequenz) [reply src/lib.rs]: yes, we changed that to required

File: src/tests.rs
Comment (shsms) [reply src/tests.rs]: So what are we doing about this?  I'd rather keep the existing name and match the behaviour of the SDK resampler.  So if `max_age` is 0, the no values are used, and if it is 1, just the data from this window are used, etc.

Comment (phillip-wenig-frequenz) [reply src/tests.rs]: okay, sounds reasonable

Comment (phillip-wenig-frequenz) [reply src/tests.rs:353]: However... It does not make sense to set it to 0. Also regarding the way it is implemented now, I suggest validating the input and not allowing 0.

Comment (phillip-wenig-frequenz) [reply src/tests.rs]: See new test with required behavior: https://github.com/frequenz-floss/frequenz-resampling-rs/pull/1/files#diff-a0262fea108af44a98cc49e5eb72641963dd816513f2d0a22eb738fcfed55ebeR174

**Comment by phillip-wenig-frequenz** [comment]
Heads up: the custom resampling-function is now Box<... + Send>, because the reporting service needs this requirement.

https://github.com/frequenz-floss/frequenz-resampling-rs/pull/1/files#diff-aa26639549dc5fd2a82138018ef592341e90d8b70a34ee45a93fe19575098702R15

File: README.md
Comment (llucax) [reply README.md]: I'm a bit confused. This seems to be talking about upsampling, right?

> Yes, we have the functionality here that the interval that is multiplied with max_age adapts to the input interval, if the input interval is larger than the set interval.
> 
> self.input_interval = Some((s.timestamp() - input_start).max(self.interval));
> 
> Is that already what you have in mind, @llucax ?

But this is about the zero `max_age`?

> There is also a test for this behavior: [a0262fea108af44a98cc49e5eb72641963dd816513f2d0a22eb738fcfed55ebeR201](https://github.com/frequenz-floss/frequenz-resampling-rs/pull/1/files#diff-a0262fea108af44a98cc49e5eb72641963dd816513f2d0a22eb738fcfed55ebeR201)

About the test, I'm not sure if I am misunderstanding the text above or the code, but that test seem to show that basically the resampling produces all samples with `None` as value when `max_age_in_intervals` is `0`. Is this correct? In the text above for me it sounds like the resampling is done with *something* (not sure what) and only then, after the resampling was done, the buffer is cleared out, which looks like the resampled samples shouldn't produce `None` in the test you linked.

The test looks OK, but I'm not sure if there is a point in allowing zero max age at all though, it ends up with `None` sample factory, right?

File: src/tests.rs
Comment (llucax) [reply src/tests.rs]: OK, it seems this changed from the last review, so probably the readme needs updating too?

File: README.md
Comment (phillip-wenig-frequenz) [reply README.md]: About upsampling: This is the new location of the test. Was quite confused too. https://github.com/frequenz-floss/frequenz-resampling-rs/pull/1/files#diff-a0262fea108af44a98cc49e5eb72641963dd816513f2d0a22eb738fcfed55ebeR321

File: src/tests.rs
Comment (cwasicki) [reply src/tests.rs]: >So if max_age is 0, the no values are used, and if it is 1, just the data from this window are used, etc.

So 0 should be forbidden then right? Would make sense to me, that `max_age` gives the number of sampling intervals with data, i.e. 1 contains only data from 1 sampling interval. 

W.r.t. naming, I understand to keep the desire to keep the existing name, but tbh I think it's not clear at all. To me `max_age` would be a `timedelta`.

Comment (phillip-wenig-frequenz) [reply src/tests.rs]: Yes, it wouldn't make sense to use `0`.

Comment (llucax) [reply src/tests.rs]: > W.r.t. naming, I understand to keep the desire to keep the existing name, but tbh I think it's not clear at all. To me `max_age` would be a `timedelta`.

That's another option, we considered it for the SDK resampler, but went for number of intervals because it is optional in the SDK, and having an absolute `max_age` didn't play nice with that (we use a default of `max_age=3`, so that would be 3 windows if you use `interval=1s` but 1 window if you use `interval=3s`, which is unexpected IMHO). Usually what you want to use `max_age` for is to set some fault tolerance, and we thought that's better expressed relative to the interval.

If it is mandatory, I guess it could be a time delta, but in practice, if you usually will end up setting the `max_age` as some multiple of the interval, it makes the code much more verbose (like `Resampler::new(interval, ResamplingFunction::Average, interval * 3, start);`

Comment (phillip-wenig-frequenz) [reply src/tests.rs]: I think this is the best solution

Comment (phillip-wenig-frequenz) [reply src/tests.rs]: The `max_age` as a multiplier has one advantage though and that's the adaptability to a slower input interval. If you set `interval` to 1s, but the data comes in every 2s, then the interval adapts to this frequency (to avoid empty resamples). This interval gets multiplied with `max_age`. 

With `max_age` being an interval itself, we can only assume that an adaptation to `1 x input_interval` is fine.

Comment (llucax) [reply src/tests.rs]: Yeah, good point. That too.

Comment (cwasicki) [reply src/tests.rs]: In my opinion it should be a multiple of the interval. My comment was just about the name of the parameter.

Comment (shsms) [reply src/tests.rs]: People are still typing `max_age`, but wasn't it already renamed to `max_age_in_invervals`?

Comment (phillip-wenig-frequenz) [reply src/tests.rs]: So, what do we do? 

1. Keeping `max_age_in_invervals` (my preference)
2. Renaming it to `reuse_previous_intervals` or something similar
3. Making it an interval

Comment (cwasicki) [reply src/tests.rs]: Also voting for 1.

Comment (llucax) [reply src/tests.rs]: :one:

**Review by shsms — APPROVED** [review]
This looks good to me, so I'm approving it. But please wait for @llucax to confirm as well before merging.

**Review by llucax — APPROVED** [review]
Pretty happy with it too as a first approach.

**Comment by phillip-wenig-frequenz** [comment]
@shsms @llucax great, thank you! Can you merge it? I do not have write access to this repository.

**Comment by shsms** [comment]
> Can you merge it? I do not have write access to this repository.

We can wait until @llucax gives you access.

**Comment by shsms** [comment]
Before you merge, please update the PR title.  It will be used as the merge commit message.

**Comment by llucax** [comment]
I think adding a description explaining what was implemented so far (for example) can be a good idea too (it will be used as the commit body).

**Comment by llucax** [comment]
@phillip-wenig-frequenz you should be able to self-merge now.

**Pull request merged**
